src/documents/models.py

```diff
@@ -267,17 +277,90 @@ def __str__(self):
         return str(created)
 
     @property
-    def source_path(self):
+    def source_filename(self):
+        if self.filename is None:
+            self.filename = self.generate_source_filename()
+
+        return self.filename
+
+    @staticmethod
+    def many_to_dictionary(field):
+        # Converts ManyToManyField to dictionary by assuming, that field
+        # entries contain an _ or - which will be used as a delimeter
+        mydictionary = dict()
+
+        for t in field.all():
+            # Find delimeter
+            delimeter = t.name.find('_')
+
+            if delimeter is -1:
+                delimeter = t.name.find('-')
+
+            if delimeter is -1:
+                continue
+
+            key = t.name[:delimeter]
+            value = t.name[delimeter+1:]
+
+            mydictionary[slugify(key)] = slugify(value)
+
+        return mydictionary
+
+    @staticmethod
+    def many_to_list(field):
+        # Converts ManyToManyField to list
+        mylist = list()
+
+        for t in field.all():
+            mylist.append(slugify(t))
+
+        return mylist
 
-        file_name = "{:07}.{}".format(self.pk, self.file_type)
+    def generate_source_filename(self):
+        # Create filename based on configured format
+        if settings.PAPERLESS_FILENAME_FORMAT is not None:
+            path = settings.PAPERLESS_FILENAME_FORMAT.format(
+                   correspondent=slugify(self.correspondent),
+                   title=slugify(self.title),
+                   created=slugify(self.created),
+                   added=slugify(self.added),
+                   tag=defaultdict(str, self.many_to_dictionary(self.tags)),
```

Comment:
While this doesn't suffer from any `KeyError`s from occuring, thanks to the default dict, it does leave weird filenames if the format specified unconditional content, e.g.:

```
PAPERLESS_FILENAME_FORMAT={correspondent}-{tag[type]}
```

Would result in a file named e.g. `MyBank--0000000.pdf` -- notice the double-dashes. Even worse: if the user specified the filename to be something like `{tag[foo]}-{tag[bar]}` it could cause filenames like `--0000000.pdf` or `-bar-0000000.pdf`.

Now while leading dashes should work on most filesystems, at least handling them with a common shell is weird because [they might be interpreted as options](https://unix.stackexchange.com/questions/1519/how-do-i-delete-a-file-whose-name-begins-with-hyphen-a-k-a-dash-or-minus).

Considering my other comment about the next line potentially causing `IndexError`s, maybe the behaviour here should be different: try to format the string as provided, catching `KeyError`s and `IndexError`s. If either occurs, fallback to the default numbers-only filename.

Reply:
Good point - fixed by defaulting to "none" which ensure never to have a - as prefix. Also, -- will be prevented unless explicitly configured by the user.